[torchrec][LocalShardsWrapper] Implement tensor padding for local shards wrapper

jeffkbkim opened the pull request:
Differential Revision: D82663766

This patch implements the constant padding functionality (aten.constant_pad_nd.default) for LocalShardsWrapper. The method applies constant padding to the local shards based on the provided padding specification.

Padding may be used to resize tensors to the same size in order to perform an all gather. LocalShardsWrapper, a wrapper around a DTensor's local_tensor, can now support padding across its local shards.

Depending on the sharding type (RW, CW), the padding on [left, right, top, bottom] directions will be either applied to the first/last shard, or all local shards.


cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci

ezyang reviewed (changes requested):
If we're doing this here, the tests need to live in this repo. But I also need some more context on this change.

jeffkbkim commented:
> @ezyang Thanks for the response. I have updated the PR description.
> 
> I do have a test file that I can move - where would you recommend the test file to be located?
> 
> I am not too familiar of this codebase and was unable to find an appropriate directory.

@ezyang I have added a new test under test/distributed/tensor/test_shards_wrapper.py

ezyang commented:
I don't know if we actually have a reviewer for this code

jeffkbkim committed:
[torchrec][LocalShardsWrapper] Implement tensor padding for local shards wrapper (#163183)

Summary:
X-link: https://github.com/pytorch/torchrec/pull/3382


This diff implements the constant padding functionality (aten.constant_pad_nd.default) for `LocalShardsWrapper`. The method applies constant padding to the local shards based on the provided padding specification.

Depending on the sharding type (RW, CW), the padding on [left, right, top, bottom] directions will be either applied to the first/last shard, or all local shards.

New unit tests cover:
- 1D (RW) top/bottom paddings
- 2D (CW) left, right, top, bottom paddings
- empty shards, number of dimensions > 2

Test Plan:
```
buck2 test fbcode//caffe2/test/distributed/tensor:shards_wrapper
2025-09-18T15:32:46.525914Z  WARN buck2_interpreter_for_build::interpreter::functions::warning: ptxas 12.8 is not available on platform platform010-aarch64-compat
2025-09-18T15:32:46.525953Z  WARN buck2_interpreter_for_build::interpreter::functions::warning: ptxas 12.8 is not available on platform platform010-compat
2025-09-18T15:32:46.525959Z  WARN buck2_interpreter_for_build::interpreter::functions::warning: ptxas 12.8 is not available on platform platform010-libcxx
Buck UI: https://www.internalfb.com/buck2/ffb34bcb-1555-4fa3-89c6-9c22d078606a
Test UI: https://www.internalfb.com/intern/testinfra/testrun/12384899087608299
Network: Up: 159MiB  Down: 13GiB  (reSessionID-f734bd3c-19ca-44c9-919f-57203ac00be8)
Loading targets.   Remaining      0/5110                                                                                                                                                                        104336 dirs read, 1265395 targets declared
Analyzing targets. Remaining      0/80346                                                                                                                                                                       3349033 actions, 4142832 artifacts declared
Executing actions. Remaining      0/521855                                                                                                                                                                      149:06:17.8s exec time total
Command: test.     Finished 14 local, 397 remote, 199840 cache (99% hit)                                                                                                                                        148:27:40.9s exec time cached (99%)                                  
Time elapsed: 8:55.5s
Tests finished: Pass 14. Fail 0. Fatal 0. Skip 0. Build failure 0
```

Differential Revision: D82663766
